samples/grpc/restaurant-drone-deliveries-service-java/src/main/java/charging/ChargingStation.java

```diff
@@ -0,0 +1,369 @@
+                  return Effect().none();
+                });
+
+    var initializedHandler =
+        newCommandHandlerBuilder()
+            .forNonNullState()
+            .onCommand(
+                Create.class,
+                create ->
+                    Effect()
+                        .reply(
+                            create.replyTo,
+                            StatusReply.error(
+                                "Got a create command, but station id "
+                                    + getReplicationContext().entityId()
+                                    + " was already created")))
+            .onCommand(StartCharging.class, this::handleStartCharging)
+            .onCommand(
+                CompleteCharging.class,
+                (state, completeCharging) -> {
+                  context.getLog().info("Drone {} completed charging", completeCharging.droneId);
+                  if (state.dronesCharging.stream()
+                      .anyMatch(
+                          chargingDrone -> chargingDrone.droneId.equals(completeCharging.droneId)))
+                    return Effect()
+                        .persist(new ChargingCompleted(completeCharging.droneId))
+                        .thenReply(completeCharging.replyTo, newState -> StatusReply.ack());
+                  else
+                    return Effect()
+                        .reply(
+                            completeCharging.replyTo,
+                            StatusReply.error(
+                                "Drone "
+                                    + completeCharging.droneId
+                                    + " is not currently charging."));
+                })
+            .onCommand(
+                GetState.class,
+                (state, getState) -> Effect().reply(getState.replyTo, StatusReply.success(state)));
+
+    return noStateHandler.orElse(initializedHandler).build();
+  }
+
+  private Effect<Event, State> handleStartCharging(State state, StartCharging startCharging) {
+    if (state.dronesCharging.stream()
+        .anyMatch(charging -> charging.droneId.equals(startCharging.droneId))) {
+      return Effect().reply(startCharging.replyTo, StatusReply.error("Drone already charging"));
+    } else if (state.dronesCharging.size() >= state.chargingSlots) {
+      var earliestFreeSlot =
+          state.dronesCharging.stream()
+              .min(Comparator.comparing(chargingDrone -> chargingDrone.chargingDone))
+              .get()
+              .chargingDone;
+      context
+          .getLog()
+          .info(
+              "Drone {} requested charging but all stations busy, earliest free slot {}",
+              startCharging.droneId,
+              earliestFreeSlot);
+      return Effect()
+          .reply(startCharging.replyTo, StatusReply.success(new AllSlotsBusy(earliestFreeSlot)));
+    } else {
+      // charge
+      var chargeCompletedBy = Instant.now().plus(FULL_CHARGE_TIME);
```

Comment:
We still estimate when done, needed because if all stations are busy we reply with a "try again at this point in time" calculated from first drone expected to have completed charging (to have some more logic than just is-list-full-then-deny)

Reply:
I see, 👍

Reply:
I'll rename to `expectedComplete` across events and commands/APIs to make it more more clear that it is a guess and not controlled by the entity itself